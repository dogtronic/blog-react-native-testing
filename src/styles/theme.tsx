import React, {ReactElement} from 'react';
import * as styledComponents from 'styled-components/native';

type colors = {
  system: {
    tokenOnWhite: string;
    tokenGreen: string;
    tokenGreen60: string;
    tokenGreen30: string;
    tokenGreen10: string;
    tabBarBorder: string;
  };
  text: {
    black: string;
    darkGrey: string;
    grey: string;
    lightGrey: string;
    highlightGrey: string;
    white: string;
  };
  status: {
    dangerous: string;
    bad: string;
    good: string;
  };
  background: {
    light: string;
    dark: string;
  };
};

declare module 'styled-components' {
  export interface DefaultTheme {
    colors: colors;
  }
}

export const theme: styledComponents.DefaultTheme = {
  colors: {
    system: {
      tokenOnWhite: '#14CA85',
      tokenGreen: '#00E69D',
      tokenGreen60: '#6FFFE3',
      tokenGreen30: '#B9FFE8',
      tokenGreen10: '#E7FFF7',
      tabBarBorder: '#E2E3E4',
    },
    text: {
      black: '#000000',
      darkGrey: '#36354A',
      grey: '#686777',
      lightGrey: '#B3B5BD',
      highlightGrey: '#D2D2D7',
      white: '#FFFFFF',
    },
    status: {
      dangerous: '#FF0000',
      bad: '#FF6B00',
      good: '#15AA2C',
    },
    background: {
      light: '#FFFFFF',
      dark: '#F8F8FA',
    },
  },
};

export type ThemeInterface = typeof theme;

const {
  default: styled,
  css,
  ThemeProvider,
} = styledComponents as styledComponents.ReactNativeThemedStyledComponentsModule<ThemeInterface>;

const withTheme = (element: ReactElement) => {
  <ThemeProvider theme={theme}>{element}</ThemeProvider>;
};

export {styled, css, ThemeProvider, withTheme};
