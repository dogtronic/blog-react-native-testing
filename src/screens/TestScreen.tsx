import React, {useState} from 'react';
import SimpleButton from '../components/SimpleButton';
import {styled} from '../styles/theme';
import {axiosCall} from '../utils/axiosCall';

const StyledSafeAreaView = styled.SafeAreaView`
  align-items: center;
  margin-horizontal: 20px;
`;
const StyledText = styled.Text`
  margin-vertical: 20px;
`;

const TestScreen: React.FC = () => {
  const [data, setData] = useState({age: 0, count: 0, name: 'noname'});

  const callAPI = () => {
    axiosCall()
      .then(response => setData(response))
      .catch(err => console.warn(err));
  };
  return (
    <StyledSafeAreaView>
      <StyledText>{`age: ${data.age} name: ${data.name}`}</StyledText>
      <SimpleButton
        label={'Call API'}
        onPress={callAPI}
        buttonTestID={'simpleButtonTestID'}
      />
    </StyledSafeAreaView>
  );
};

export default TestScreen;
