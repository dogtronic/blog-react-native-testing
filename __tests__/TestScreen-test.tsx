import React from 'react';
import {fireEvent, render, waitFor} from '@testing-library/react-native';
import TestScreen from '../src/screens/TestScreen';
import {withTheme} from '../src/styles/theme';

describe('TestScreen', () => {
  test('renders correctly', () => {
    const testScreenRender = render(withTheme(<TestScreen />));
    expect(testScreenRender.toJSON()).toMatchSnapshot();
  });
  test('renders correctly after user button press', async () => {
    const testScreenRender = render(withTheme(<TestScreen />));
    const axiosCall = require('../src/utils/axiosCall');
    const mockAxiosCall = jest
      .spyOn(axiosCall, 'axiosCall')
      .mockResolvedValue({age: 70, count: 233482, name: 'michael'});

    fireEvent(testScreenRender.getByTestId('simpleButtonTestID'), 'press');

    await waitFor(() => expect(mockAxiosCall).toHaveBeenCalledTimes(1)).then(
      () => expect(testScreenRender.toJSON()).toMatchSnapshot(),
    );
  });
});
