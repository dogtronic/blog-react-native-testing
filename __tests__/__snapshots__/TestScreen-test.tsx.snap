// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TestScreen renders correctly 1`] = `
<RCTSafeAreaView
  style={
    Array [
      Object {
        "alignItems": "center",
        "marginHorizontal": 20,
      },
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "marginVertical": 20,
        },
      ]
    }
  >
    age: 0 name: noname
  </Text>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#F8F8FA",
        "flexDirection": "row",
        "height": 56,
        "justifyContent": "center",
        "marginTop": 17,
        "opacity": 1,
        "width": "100%",
      }
    }
    testID="simpleButtonTestID"
  >
    <Text
      style={
        Array [
          Object {
            "color": "#B3B5BD",
            "fontFamily": "RobotoMono-Medium",
            "fontSize": 16,
            "lineHeight": 21,
          },
        ]
      }
    >
      Call API
    </Text>
  </View>
</RCTSafeAreaView>
`;

exports[`TestScreen renders correctly after user button press 1`] = `
<RCTSafeAreaView
  style={
    Array [
      Object {
        "alignItems": "center",
        "marginHorizontal": 20,
      },
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "marginVertical": 20,
        },
      ]
    }
  >
    age: 70 name: michael
  </Text>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#F8F8FA",
        "flexDirection": "row",
        "height": 56,
        "justifyContent": "center",
        "marginTop": 17,
        "opacity": 1,
        "width": "100%",
      }
    }
    testID="simpleButtonTestID"
  >
    <Text
      style={
        Array [
          Object {
            "color": "#B3B5BD",
            "fontFamily": "RobotoMono-Medium",
            "fontSize": 16,
            "lineHeight": 21,
          },
        ]
      }
    >
      Call API
    </Text>
  </View>
</RCTSafeAreaView>
`;
